### PR TITLE
[P4-1631] Streamline Pipeline "Part I - Divide and Conquer"

### PR DESCRIPTION
**Pull request by mafonso**
### Jira link

[[P4-1631]](https://dsdmoj.atlassian.net/browse/P4-1631)

### What?

I have added/removed/altered:

- [ ] Introduced [executors](https://circleci.com/docs/2.0/executor-intro/) instead of image references.  
- [ ] Split the individual tests and made them use parallel workers.

### Why?

I am doing this because:
- It makes it easier and cleaner to define the CI job's environments. 
- It speeds up tests a lot! 

image

### Deployment risks (optional)

This PR just affects the workflow up to the PR stage, the rest keep fairly the same. The deployment approach will be changed in part-2, making this one lower risk and simpler to review.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -210,87 +251,96 @@ commands:
...
+    parallelism: 8
+    steps:
+      - checkout
+      - build-base
+      - *wait-for-db
+      - *create-db
+      - *attach-tmp-workspace
+      - *rspec
+
+  rubocop_tests:
```
Comment (willfish) [inline .circleci/config.yml:291]:
```suggestion
  linters:
```

Comment (willfish) [reply .circleci/config.yml:291]: I suspect we'll want to extend these to include some security checks with brakeman, etc

File: .circleci/config.yml [hunk 2]
```diff
@@ -301,18 +351,29 @@ workflows:
...
           <<: *all_tags
       - api_docs:
           <<: *all_tags
+          requires:
+            - setup_test_environment
+      - rspec_tests:
+          <<: *all_tags
+          requires:
+            - setup_test_environment
+      - rubocop_tests:
```
Comment (willfish) [inline .circleci/config.yml:364]:
```suggestion
      - linters:
```

Comment (willfish) [reply .circleci/config.yml:364]: I **love** the rubocop separation. It'll give us much better feedback about what category of failures are happening.

File: .circleci/config.yml [hunk 3]
```diff
@@ -301,18 +351,29 @@ workflows:
...
       - build_dev:
           <<: *only_master_and_create_dev
           requires:
             - api_docs
       - deploy_dev:
           <<: *only_master_and_create_dev
           requires:
-            - test
+            - rspec_tests
+            - rubocop_tests
```
Comment (willfish) [inline .circleci/config.yml:376]:
```suggestion
            - linters
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -321,7 +382,8 @@ workflows:
       - deploy_staging:
           <<: *only_master
           requires:
-            - test
+            - rspec_tests
+            - rubocop_tests
```
Comment (willfish) [inline .circleci/config.yml:386]:
```suggestion
            - linters
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -330,7 +392,8 @@ workflows:
       - deploy_preprod:
           <<: *only_deploy_tags
           requires:
-            - test
+            - rspec_tests
+            - rubocop_tests
```
Comment (willfish) [inline .circleci/config.yml:396]:
```suggestion
            - linters
```

File: .circleci/config.yml [hunk 6]
```diff
@@ -340,12 +403,14 @@ workflows:
           <<: *only_deploy_tags
           type: approval
           requires:
-            - test
+            - rspec_tests
+            - rubocop_tests
```
Comment (willfish) [inline .circleci/config.yml:407]:
```suggestion
            - linters
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -340,12 +403,14 @@ workflows:
...
             - build_production
       - notify_of_approval:
           <<: *only_deploy_tags
           requires:
-            - test
+            - rspec_tests
+            - rubocop_tests
```
Comment (willfish) [inline .circleci/config.yml:413]:
```suggestion
            - linters
```

File: .circleci/config.yml (hunk 1 above)
Comment (mafonso) [reply .circleci/config.yml:291]: Yeah! Good point.

**Pull request merged**